### PR TITLE
Add methods to wait for flushing

### PR DESCRIPTION
**Pull request by Raynos**
This allows for better functionality around waiting for SQS
server to do things.

Now we can wait for messages to be received and wait for messages
to have been flushed (aka deleted)

### CODE REVIEW

File: lib/server.js [hunk 1]
```diff
@@ -9,24 +9,19 @@ class SimplyImitatedSQSHttpServer {
     this.server = null
     this.hostPort = null
     this.waiters = []
+    this.pendingItems = 0
+    this.sendItems = 0
```
Comment (juliangruber) [inline lib/server.js]:
```suggestion
    this.sentItems = 0
```

It looks like this would be past tense

File: lib/server.js [hunk 2]
```diff
@@ -163,13 +160,46 @@ class SimplyImitatedSQSHttpServer {
...
-    this.waiters.push({ count: count, cb: cb })
+    this.waiters.push({ type: 'send', count: count, cb: cb })
+  }
+
+  waitForFlush (cb) {
+    if (this.pendingItems === 0) {
+      return process.nextTick(cb)
+    }
+
+    this.waiters.push({ type: 'pending', count: 0, cb: cb })
```
Comment (juliangruber) [inline lib/server.js:202]: What about instead of manually tracking the subscribers, we inherit from `EventEmitter` and reuse that logic?

**Review by juliangruber — CHANGES_REQUESTED** [review]
this still needs documentation

File: lib/server.js (hunk 2 above)
Comment (Raynos) [reply lib/server.js:202]: because of managing the `count` logic using the event emitter doesn't match cleanly.

File: lib/server.js (hunk 1 above)
Comment (Raynos) [reply lib/server.js]: willfix.

File: lib/server.js (hunk 2 above)
Comment (juliangruber) [reply lib/server.js:202]: would https://nodejs.org/api/events.html#events_emitter_listenercount_eventname help?

Comment (juliangruber) [reply lib/server.js:202]: so to get the count of waiters with type send, you'd do `this.listenerCound('send')`

Comment (Raynos) [reply lib/server.js:202]: count is how many messages to wait for; not related to how many waiters or listeners there are.

File: README.md [hunk 3]
```diff
@@ -47,3 +47,37 @@ myServer.run({
...
+### `const server = new SQSServer()`
+
+Create a fake SQS server
+
+### `server.run(opts, cb)`
+
+ - `opts.port` ; defaults to 0
+ - `opts.host` ; defaults to `localhost`
+
+Starts the server. cb get's called once listening on a port.
```
Comment (juliangruber) [inline README.md]:
```suggestion
Starts the server. `cb` get's called once listening on a port.
```

File: README.md [hunk 4]
```diff
@@ -47,3 +47,37 @@ myServer.run({
...
+
+### `server.getQueue()`
+
+Returns the current array of items queued in SQS
```
Comment (juliangruber) [inline README.md]: Of what shape are items?

File: README.md [hunk 5]
```diff
@@ -47,3 +47,37 @@ myServer.run({
...
+
+### `server.waitForMessages(count, listener)`
+
+Get notified once N messages have been send to this fake SQS
```
Comment (juliangruber) [inline README.md]: `listener` is a callback, and not an eventemitter listener that can be called multiple times, right?

File: README.md [hunk 6]
```diff
@@ -47,3 +47,37 @@ myServer.run({
...
+
+### `server.waitForFlush(listener)`
+
+Get notified when the number of pending messages in the SQS
+queue is zero.
```
Comment (juliangruber) [inline README.md:78]: same as above. Is this called once or every time?

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (Raynos) [reply README.md]: it's not a `cb` because there's never an error. It's a `listener` that gets called once.

File: lib/server.js (hunk 2 above)
Comment (juliangruber) [reply lib/server.js:202]: oooh gotcha!

Comment (juliangruber) [reply lib/server.js:202]: Would still be possible with event emitters but not really an improvement any more:

```js
const onSend = () => cb()
onSend.count = count
this.on('send', onSend)

// after send
const listeners = this.listeners('send')
for (const listener of listeners) {
  listener.count--
}
```

File: README.md [hunk 7]
```diff
@@ -47,3 +47,42 @@ myServer.run({
...
+Starts the server. `cb` get's called once listening on a port.
+
+### `server.getQueue()`
+
+Returns the current array of items queued in SQS. These are shaped
+like aws SQS objects.
+
+### `server.waitForMessages(count, listener)`
+
+Get notified once N messages have been send to this fake SQS.
```
Comment (juliangruber) [inline README.md]:
```suggestion
Get notified once N messages have in total been sent to this fake SQS.
```

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (juliangruber) [reply README.md]: The API was confusing to me, I thought it would call back once the next `N` messages have been sent.

**Review by juliangruber — APPROVED** [review]

**Pull request merged**